### ISSUE TITLE
Creating a temporary array from an iterator within an expression

### ISSUE THREAD

**Issue by ben-albrecht** [chapel-lang/chapel#14364]
Chapel's design philosophy prohibits the creation of temporary arrays implicitly. As a result, working with iterator expressions in Chapel can sometimes be verbose, due to the need to explicitly create temporary arrays.

*Example 1:*

```chpl
var s = someString.split().size;
```

must be:

```chpl
var tmp = someString.split();
var s = tmp.size;
```

*Example 2:*

```chpl
// A and B are arrays, f() expects an array, A-B is a promoted iterator expression
var result = f(A-B);
```

must be:

```chpl
var tmp = A - B;
var result = f(tmp);
```

It would be nice to have a way to explicitly opt into creation of temporary array within an expression, for example, using `toArray()` as a straw-person method name:

```chpl
var s = someString.split().toArray().size;
```

```chpl
var result = f((A-B).toArray());
```

Issues that would be resolved by this feature:

- https://github.com/chapel-lang/chapel/issues/7936
- https://github.com/chapel-lang/chapel/issues/7982 (partially)

**Comment by bradcray** [comment]
This request resonates with me, though I'd probably express it using cast syntax rather than a method call.

That said, for example 1, I would've thought that you'd want a way to query the size of an iterator rather than the ability to request a temporary array.  E.g., for many iterator expressions, the size could be determined without manifesting the iterator as an array.

For example 2, I assume that `f()` is intended to be a function that was written to require an array argument?

**Comment by ben-albrecht** [comment]
> This request resonates with me, though I'd probably express it using cast syntax rather than a method call.

Yeah, I'm not crazy about the proposed syntax in this issue - just wanted to jot something down for now. A cast syntax would be nice, but would would that look like for casting to an array?

> That said, for example 1, I would've thought that you'd want a way to query the size of an iterator rather than the ability to request a temporary array. E.g., for many iterator expressions, the size could be determined without manifesting the iterator as an array.

Ideally, yes. Both features would be valuable.

> For example 2, I assume that f() is intended to be a function that was written to require an array argument?

That's correct. This example was a lazy distillation from #7936.

**Comment by bradcray** [comment]
> A cast syntax would be nice, but would would that look like for casting to an array?

I was imagining we'd ultimately want to support a number of variations like:

```chapel
(A-B): A.type
(A-B): [A.domain] A.eltType
(A-B): [1..n] real
(A-B): [1..] real
(A-B): [] real
(A-B): []
```

where I suspect the last one is probably problematic from a parsing standpoint.

**Comment by bradcray** [comment]
> This example was a lazy distillation from #7936.

I updated the comment to address the laziness / potential for confusion.

**Comment by ben-albrecht** [comment]
> I updated the comment to address the laziness / potential for confusion.

Thanks @bradcray!

I like the array casting syntax significantly more than the original method style proposed.

**Comment by mppf** [comment]
Note that #16007 is a recent case running into this issue (and arguably proposing a different solution).
